README.md

# bouncing_character
[![pub package](https://img.shields.io/pub/v/bouncing_character.svg)](https://pub.dartlang.org/packages/bouncing_character)

**A flutter plugin to make elegant illustration of bouncing character on flutter web**


[README](https://github.com/iamabhishek229313/bouncing_character/blob/master/README.md)

**Github**: https://github.com/iamabhishek229313/bouncing_character

[Update log](/CHANGELOG.md)

## Usage:

### Add dependency：
Please check the latest version before installation.
```
dependencies:
  flutter:
    sdk: flutter
  # add bouncing_character
  bouncing_character: ^1.0.0
```

### Add the following imports to your Dart code:
```
import 'package:bouncing_character/bouncing_character.dart';
```
### API
**Too simple example**
```dart
    Container(
        width : 600.0,
        height : 300.0,
        child : BouncingCharacter(
            text : "I'm a developer" , // it is @required
            duration : Duration(milliseconds : 200) , // it's too
        )       
    )
```
**All functionaries**
```dart
    
  // Initializing some custom attributes .
  const BgColor = Color.fromRGBO(27, 26, 31, 1.0);
  const TextStyle customStyle = TextStyle(fontSize: 44.0, fontFamily: 'BioRhyme', color: Colors.grey);
  const Duration customDuration = Duration(milliseconds: 200);
  Matrix4 customTransform = Matrix4.identity()..translate(0, -10, 0);   
  
   BouncingCharacter(
        text: "I'm a developer",
        duration: customDuration,
        style: customStyle,
        onHoverTranslate: customTransform,
    )
```
##Example 
**main.dart**
```dart
import 'package:bouncing_character_example/home_page.dart';
import 'package:flutter/material.dart';
import 'package:bouncing_character/bouncing_character.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: HomePage()
    );
  }
}
```

**home_page.dart**
```dart
import 'package:bouncing_character/bouncing_character.dart';
import 'package:flutter/material.dart';

// Initializing some custom attributes .

const BgColor = Color.fromRGBO(27, 26, 31, 1.0);
const TextStyle customStyle = TextStyle(fontSize: 44.0, fontFamily: 'BioRhyme', color: Colors.grey);
const Duration customDuration = Duration(milliseconds: 200);
Matrix4 customTransform = Matrix4.identity()..translate(0, -10, 0);


class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: BgColor,
        body: Center(
            child: BouncingCharacter(
              text: "I'm an developer",
              duration: customDuration,
              style: customStyle,
              onHoverTranslate: customTransform,
            )));
  }
}
```
|                                                     To give you an idea ...                                                      |
|:----------------------------------------------------------------------------------------------------------------------:|
![](https://github.com/iamabhishek229313/bouncing_character/blob/master/illustration/illustration_gif.gif?raw=true) 

## Created and Maintained by. 
[Abhishek Kumar](https://github.com/iamabhishek229313) 



CHANGELOG.md
## [version] - 1.1.0
* Second release

## [1.1.0] - 2020-06-11
 
* Use this package to make an awesome boucing animation on each of the characters on hover (FLutter web) .


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:bouncing_character_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/bouncing_character_test.dart
//import 'package:flutter/services.dart';
//import 'package:flutter_test/flutter_test.dart';
//import 'package:bouncing_character/bouncing_character.dart';
//
//void main() {
//  const MethodChannel channel = MethodChannel('bouncing_character');
//
//  TestWidgetsFlutterBinding.ensureInitialized();
//
//  setUp(() {
//    channel.setMockMethodCallHandler((MethodCall methodCall) async {
//      return '42';
//    });
//  });
//
//  tearDown(() {
//    channel.setMockMethodCallHandler(null);
//  });
//
//  test('getPlatformVersion', () async {
//    expect(await BouncingCharacter.platformVersion, '42');
//  });
//}


lib/bouncing_character.dart
library bouncing_character ;

import 'package:flutter/material.dart';

class BouncingCharacter extends StatefulWidget {
  final String text ;
  final TextStyle style ;
  final Duration duration ;
  final Matrix4 onHoverTranslate ;
  final WrapAlignment wrapAlignment ;
  const BouncingCharacter({Key key,@required this.text , @required this.duration , this.style , this.onHoverTranslate, this.wrapAlignment}) : assert(
  text != null && duration != null
  ) ;

  @override
  _BouncingCharacterState createState() => _BouncingCharacterState();
}

class _BouncingCharacterState extends State<BouncingCharacter> {
  @override
  Widget build(BuildContext context) {
    List<String> words = widget.text.split(' ') ;
    print(words) ;
    return Wrap(
        runAlignment: widget.wrapAlignment == null ?  WrapAlignment.center : widget.wrapAlignment,
        children: words.map((e) => EachString(
          text: e,
          style: widget.style,
          duration: widget.duration,
          onHoverTranslate: widget.onHoverTranslate,
        ))
            .toList());
  }
}

class EachString extends StatefulWidget {
  String text;
  final TextStyle style ;
  final Duration duration ;
  final Matrix4 onHoverTranslate ;
  EachString({Key key, @required this.text ,@required this.style ,@required this.duration,@required this.onHoverTranslate}) : super(key: key);

  @override
  _EachStringState createState() => _EachStringState();
}

class _EachStringState extends State<EachString> {
  @override
  Widget build(BuildContext context) {
    widget.text = " " + widget.text ;
    return Row(
        mainAxisSize: MainAxisSize.min,
        children: List.generate(
          widget.text.length,
              (index) => TransalteContainer(
            duration: widget.duration,
            onHoverTranslate: widget.onHoverTranslate,
            child: Text(
              widget.text[index],
              style: widget.style == null ? null : widget.style,
            ),
          ),
        ));
  }
}

class TransalteContainer extends StatefulWidget {
  final Widget child;
  final Duration duration ;
  final Matrix4 onHoverTranslate ;
  const TransalteContainer({Key key, @required this.child ,@required this.duration,this.onHoverTranslate}) : super(key: key);

  @override
  _TransalteContainerState createState() => _TransalteContainerState();
}

class _TransalteContainerState extends State<TransalteContainer> {

  bool _hovering = false;

  @override
  Widget build(BuildContext context) {
    final onHoverTranslate = widget.onHoverTranslate == null ? (Matrix4.identity()..translate(0, -10, 0)) : widget.onHoverTranslate;
    final onNonHoverTranslate = Matrix4.identity()..translate(0, 0, 0);
    return MouseRegion(
        onEnter: (e) => _mouseEnter(true),
        onExit: (e) => _mouseEnter(false),
        child: AnimatedContainer(
          duration: widget.duration,
          child: widget.child,
          transform: _hovering ? onHoverTranslate : onNonHoverTranslate,
        ));
  }

  _mouseEnter(bool val) {
    setState(() {
      _hovering = val;
    });
  }
}


android/src/main/java/com/example/bouncing_character/BouncingCharacterPlugin.java
package com.example.bouncing_character;

import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** BouncingCharacterPlugin */
public class BouncingCharacterPlugin implements FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "bouncing_character");
    channel.setMethodCallHandler(this);
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "bouncing_character");
    channel.setMethodCallHandler(new BouncingCharacterPlugin());
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else {
      result.notImplemented();
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}


example/lib/home_page.dart
import 'package:bouncing_character/bouncing_character.dart';
import 'package:flutter/material.dart';

// Initializing some custom attributes .

const BgColor = Color.fromRGBO(27, 26, 31, 1.0);
const TextStyle customStyle = TextStyle(fontSize: 44.0, fontFamily: 'BioRhyme', color: Colors.grey);
const Duration customDuration = Duration(milliseconds: 200);
Matrix4 customTransform = Matrix4.identity()..translate(0, -10, 0);


class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
        backgroundColor: BgColor,
        body: Center(
            child: BouncingCharacter(
              text: "I'm a developer",
              duration: customDuration,
              style: customStyle,
              onHoverTranslate: customTransform,
            )));
  }
}
